File: src/arm.ts
import { validateArmResetInput } from './validate.js';

export interface ArmResetInput {
  previousRatePercent: number;
  initialRatePercent: number;
  indexRatePercent: number;
  marginPercent: number;
  initialCapPercent?: number;
  periodicCapPercent?: number;
  lifetimeCapPercent?: number;
  isFirstReset: boolean;
}

export interface ArmResetResult {
  fullyIndexedRatePercent: number;
  cappedRatePercent: number;
}

/**
 * Pure rate-cap calculator. Plug cappedRatePercent into a normal renewal segment's
 * annualInterestRatePercent (the existing "variable rate change" mechanism) — call
 * calculateMonthlyPayment(balance, cappedRatePercent, remainingMonths) directly for
 * the resulting payment; this module doesn't duplicate that.
 */
export function calculateArmResetRate(input: ArmResetInput): ArmResetResult {
  validateArmResetInput(input);

  const fullyIndexedRatePercent = input.indexRatePercent + input.marginPercent;
  let capped = fullyIndexedRatePercent;

  const perResetCap = input.isFirstReset ? input.initialCapPercent : input.periodicCapPercent;
  if (perResetCap !== undefined) {
    capped = Math.min(capped, input.previousRatePercent + perResetCap);
  }
  if (input.lifetimeCapPercent !== undefined) {
    capped = Math.min(capped, input.initialRatePercent + input.lifetimeCapPercent);
  }
  capped = Math.max(capped, 0);

  return { fullyIndexedRatePercent, cappedRatePercent: capped };
}

File: COB-py/cob_calculator/arm.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from .validate import validate_arm_reset_input


@dataclass
class ArmResetInput:
    previous_rate_percent: float
    initial_rate_percent: float
    index_rate_percent: float
    margin_percent: float
    is_first_reset: bool
    initial_cap_percent: Optional[float] = None
    periodic_cap_percent: Optional[float] = None
    lifetime_cap_percent: Optional[float] = None


@dataclass
class ArmResetResult:
    fully_indexed_rate_percent: float
    capped_rate_percent: float


def calculate_arm_reset_rate(input: ArmResetInput) -> ArmResetResult:
    """Pure rate-cap calculator. Plug capped_rate_percent into a normal renewal
    segment's annual_interest_rate_percent (the existing "variable rate change"
    mechanism) -- call calculate_monthly_payment(balance, capped_rate_percent,
    remaining_months) directly for the resulting payment; this module doesn't
    duplicate that."""
    validate_arm_reset_input(input)

    fully_indexed_rate_percent = input.index_rate_percent + input.margin_percent
    capped = fully_indexed_rate_percent

    per_reset_cap = input.initial_cap_percent if input.is_first_reset else input.periodic_cap_percent
    if per_reset_cap is not None:
        capped = min(capped, input.previous_rate_percent + per_reset_cap)
    if input.lifetime_cap_percent is not None:
        capped = min(capped, input.initial_rate_percent + input.lifetime_cap_percent)
    capped = max(capped, 0)

    return ArmResetResult(fully_indexed_rate_percent=fully_indexed_rate_percent, capped_rate_percent=capped)
